server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java

```diff
@@ -274,18 +300,31 @@ public Author author() {
     }
 
     @Override
-    public MetaRepository metaRepo() {
-        MetaRepository metaRepo = this.metaRepo.get();
-        if (metaRepo != null) {
-            return metaRepo;
+    public MetaRepository resetMetaRepository() {
+        final Repository repository = repos.get(REPO_DOGMA);
+        final CompletableFuture<Entry<JsonNode>> future = repository.getOrNull(Revision.HEAD, Query.ofJson(
+                META_TO_DOGMA_MIGRATED));
+        try {
+            // Will be executed by the ZooKeeper command executor during migration.
+            final Entry<JsonNode> entry = future.get(10, TimeUnit.SECONDS);
+            return setMetaRepository(entry != null);
+        } catch (Exception e) {
+            throw new IllegalStateException("failed to get the migration entry in 10 seconds. ", e);
         }
+    }
 
-        metaRepo = new DefaultMetaRepository(repos.get(REPO_META));
-        if (this.metaRepo.compareAndSet(null, metaRepo)) {
-            return metaRepo;
-        } else {
-            return this.metaRepo.get();
-        }
+    private MetaRepository setMetaRepository(boolean useDogmaRepoAsMetaRepo) {
+        final String repoName = useDogmaRepoAsMetaRepo ? REPO_DOGMA : REPO_META;
+        final DefaultMetaRepository metaRepo = new DefaultMetaRepository(repos.get(repoName));
+        this.metaRepo = metaRepo;
+        return metaRepo;
+    }
+
+    @Override
+    public MetaRepository metaRepo() {
```

Comment:
Note: I'm going to assume the returned value of this method isn't cached anywhere so that a stale version of `MetaRepository` isn't read

Reply:
That is true. Thanks for double-checking. 😉